content/microservices-interview-questions/2-service-discovery.md

```diff
@@ -1,37 +1,83 @@
+  - In a modern, cloud‑based microservices application, however, this is a much more difficult problem to solve.
+  - Service instances have dynamically assigned network locations. 
+  - Moreover, the set of service instances changes dynamically because of autoscaling, failures, and upgrades. Consequently,your client code needs to use a more elaborate         service discovery mechanism.
+  - Service Discovery solves this problem.
 
 ---
 
 ## Question
-# What is Spring?
+# What are types of patterns/types available in Service Discovery
 
 --
 
 # What is the interviewer looking for?
-- One
-- Two
-- Three
+- If you are aware of this concept on different patterns/types of Service Discovery.
+- Which pattern you used and why.
+- Pros and Cons of each pattern.
 
 --
 
 # Answer
-- One
-- Two
-- Three
+- There are two main service discovery patterns: client‑side discovery and server‑side discovery.
+- The Client‑Side Discovery Pattern
+    - Client is responsible for determining the network locations of available service instances and load balancing requests across them.
+    - The client queries a service registry, which is a database of available service instances.
+    - The client then uses a load‑balancing algorithm to select one of the available service instances and makes a request.
+    - Netflix OSS provides a great example of the client‑side discovery pattern.
+    - Netflix Eureka is a service registry.
+    - Netflix Ribbon is an IPC client that works with Eureka to load balance requests across the available service instances.
+- Pros
+   - Since the client knows about the available services instances, it can make intelligent, application‑specific load‑balancing decisions such as using hashing consistently.
+- Cons
+   - One significant drawback of this pattern is that it couples the client with the service registry. You must implement client‑side service discovery logic for each              programming language and framework used by your service clients. 
+ - The Server‑Side Discovery Pattern 
+   -  Here the client service/application does not make a request to service registry.
+   -  Instead The client makes a request to a service via a load balancer.
+   -  The load balancer queries the service registry and routes each request to an available service instance.
+   -  The AWS Elastic Load Balancer (ELB) is an example of a server-side discovery router.
+   -  Kubernetes ectd is also one example.
+- Pros
+   - One great benefit of this pattern is that details of discovery are abstracted away from the client.
+   - This eliminates the need to implement discovery logic for each programming language and framework used by your service clients.
+- Cons
+    - Unless the load balancer is provided by the deployment environment, it is yet another highly available system component that you need to set up and manage. 
```

Comment:
Recommend breaking the "What are the two patterns?" and "Pros and cons" as two separate questions.

Reply:
Thanks for suggestion :)  will break it down